[INKLB-000] utils: base service chassiss: add internal sessions manager

sonofR9 opened the pull request:
https://trello.com/c/0lALCtXc/27-b-000-service-chassis-менеджер-сессий

sonofR9 committed:
Rename to InternalSessionsManager

It will be implemented in base class, because there should be realy only one implementation

sonofR9 committed:
Copy and move constructors are deleted

Because manager is injected into sessions:
sessions manage their lifetime by themself (as in beast examples)

sonofR9 committed:
InternalSessionsManager implemented

sonofR9 committed:
From raw pointers to weak_ptr

sonofR9 committed:
Fix typo

sonofR9 committed:
Add to cmake

Gallaann commented on back/utils/base_service_chassiss/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,125 @@
+    const Endpoint& endpoint = sessionPtr->GetClientEndpoint();
+    if (!m_sessions.contains(endpoint))
+    {
+        std::string error{};
+        std::ostringstream ss{error};
+        ss << "Manager already contains enpoint '" << endpoint.address << ':' << endpoint.port
+           << "'";
+        throw std::logic_error(error);
```
Разве не ss должен пробрасыватсья в исключение? error же - пустая строка, а вот описание ошибки в ss уже

Gallaann replied:
Разве не ss должен пробрасыватсья в исключение? error же - пустая строка, а вот описание ошибки в ss уже

Gallaann commented on back/utils/base_service_chassiss/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,125 @@
+        m_endpointByDescriptor.erase(descriptor);
+        m_DescriptorByEndpoint.erase(endpoint);
+
+        auto& docs = m_docsByUser[descriptor.login];
+        docs.erase(std::remove(docs.begin(), docs.end(), descriptor.documentId), docs.end());
+
+        auto& users = m_usersByDoc[descriptor.documentId];
+        users.erase(std::remove(users.begin(), users.end(), descriptor.login), users.end());
```
Непонятная конструкция...
Убрал всех documentId/login из мапы std::remove'ом, а потом что?

Gallaann reviewed (commented):
Посмотрел

sonofR9 committed:
Fix: stringstream.str(), std::erase

Before was vector.erase(std::remove)
Thanks for the review!

sonofR9 committed:
Add namespace

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
INKLB-001 Sessions manager uses weak_ptr to session

It should not own session
When session is called, it should be shared_ptr => it wouldn't be deleted

sonofR9 committed:
FIX std::weak_ptr<IServiceSession>

sonofR9 committed:
INKLB-001 move sessions manager

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
FIX after merge

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
INKLB-001 FIX includes

sonofR9 committed:
INKLB-001 cmake

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
INKLB-001 add test

sonofR9 committed:
INKLB-001 Add operator== to descriptor (for hash)

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
INKLB-001 FIX delete nullptr, throw if add duplicate

Gallaann commented on backend/utils/base_service_chassis/internal_sessions_manager/include/internal_sessions_manager.h:
```diff
@@ -33,28 +38,29 @@ struct std::hash<inklink::base_service_chassis::DocSessionDescriptor>
 
-    virtual void AddSession(const DocSessionDescriptor&, std::weak_ptr<IServiceSession>) = 0;
-    virtual void RemoveSession(const DocSessionDescriptor&) = 0;
-    virtual void RemoveSession(IServiceSession*) = 0;
+    virtual void AddSession(const DocSessionDescriptor&, std::weak_ptr<IServiceSession>);
+    virtual void RemoveSession(const DocSessionDescriptor&);
+    virtual void RemoveSession(IServiceSession*);
 
-    virtual std::weak_ptr<IServiceSession> GetSession(const DocSessionDescriptor&) = 0;
-    virtual std::vector<std::weak_ptr<IServiceSession>> GetSessionsByDocument(const std::string&) = 0;
-    virtual std::vector<std::weak_ptr<IServiceSession>> GetSessionsByUser(const std::string&) = 0;
-    virtual std::weak_ptr<IServiceSession> GetSession(const Endpoint&) = 0;
+    virtual std::weak_ptr<IServiceSession> GetSession(const DocSessionDescriptor&);
+    virtual std::vector<std::weak_ptr<IServiceSession>> GetSessionsByDocument(const std::string&);
+    virtual std::vector<std::weak_ptr<IServiceSession>> GetSessionsByUser(const std::string&);
+    virtual std::weak_ptr<IServiceSession> GetSession(const Endpoint&);
 
-    virtual DocSessionDescriptor GetDescriptor(const Endpoint&) = 0;
+    virtual DocSessionDescriptor GetDescriptor(const Endpoint&);
```
С чем это связано?

sonofR9 replied:
Скорее всего, класс не будет и не должен иметь других реализаций, так что он остался виртуальным только чтобы можно было заменить на stub и для тестирования

Gallaann commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+
+namespace inklink::base_service_chassis
+{
+void InternalSessionsManager::AddSession(const DocSessionDescriptor& descriptor, std::weak_ptr<IServiceSession> session)
+{
+    const std::shared_ptr<IServiceSession> sessionPtr{session};
```
Для чего weak указатель делать shared?

Gallaann reviewed (approved):
Немного не понял

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/include/internal_sessions_manager.h:
```diff
@@ -33,28 +38,29 @@ struct std::hash<inklink::base_service_chassis::DocSessionDescriptor>
-    virtual void RemoveSession(IServiceSession*) = 0;
+    virtual void AddSession(const DocSessionDescriptor&, std::weak_ptr<IServiceSession>);
+    virtual void RemoveSession(const DocSessionDescriptor&);
+    virtual void RemoveSession(IServiceSession*);
```
Почему указатель?

sonofR9 replied:
Обычно они будут по указателям (умным зачастую) храниться, поэтому решил, что так извне будет удобнее пользоваться.
Если я не прав и всё равно удобнее по ссылке, то исправлю

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/include/internal_sessions_manager.h:
```diff
@@ -33,28 +38,29 @@ struct std::hash<inklink::base_service_chassis::DocSessionDescriptor>
-    virtual std::vector<std::weak_ptr<IServiceSession>> GetSessionsByUser(const std::string&) = 0;
-    virtual std::weak_ptr<IServiceSession> GetSession(const Endpoint&) = 0;
+    virtual std::weak_ptr<IServiceSession> GetSession(const DocSessionDescriptor&);
+    virtual std::vector<std::weak_ptr<IServiceSession>> GetSessionsByDocument(const std::string&);
+    virtual std::vector<std::weak_ptr<IServiceSession>> GetSessionsByUser(const std::string&);
```
Непонятна семантика аргументов.

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+#include "internal_sessions_manager.h"
+
+#include <algorithm>
+#include <iservice_session.h>
```
Кастомные библиотеки отдельно от стандартных И идут раньше.

Бтв, почему вообще подключается так, а не через кавычки? Это другая библиотека?

sonofR9 replied:
Да, service_session в server_network библиотеке

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+        throw std::logic_error(ss.str());
+    }
+
+    m_sessions[endpoint] = session;
+    m_endpointByDescriptor[descriptor] = endpoint;
+    m_DescriptorByEndpoint[endpoint] = descriptor;
```
Кажется, что тут лучше использовать `try_emplace` с последующей проверкой.

sonofR9 replied:
До этого contains стоит
Проще сначала проверить, а потом уже изменения вносить, чтобы была строгая гарантия исключений. И она здесь нужна.
(добавил ещё проверку m_endpointByDescriptor.contains(descriptor) )
P.S. при отсутствии дескриптора и endpoint (в m_sessions) гарантировано, что всего остального тоже нет
И дважды с одного эндпоинта подключиться невозможно (так, чтобы старый был активен и находился здесь)

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+    m_endpointByDescriptor[descriptor] = endpoint;
+    m_DescriptorByEndpoint[endpoint] = descriptor;
+
+    m_docsByUser[descriptor.login].push_back(descriptor.documentId);
+    m_usersByDoc[descriptor.documentId].push_back(descriptor.login);
```
В таком формате тут может быть вызван дефолтный конструктор. Это ожидаемое поведение? Если нет, то лучше сначала искать и обрабатывать ошибки.

sonofR9 replied:
Да, так и надо
Правда, об ошибках в логине и documentId не думал, исправлю

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+
+void InternalSessionsManager::RemoveSession(const DocSessionDescriptor& descriptor)
+{
+    if (m_endpointByDescriptor.contains(descriptor))
```
Инвертировать условие.

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+{
+    if (m_endpointByDescriptor.contains(descriptor))
+    {
+        const Endpoint& endpoint = m_endpointByDescriptor[descriptor];
```
`const auto&`

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+{
+    if (m_endpointByDescriptor.contains(descriptor))
+    {
+        const Endpoint& endpoint = m_endpointByDescriptor[descriptor];
```
Всегда ли обязательно есть соответствующий endpoint? Опять может вызываться дефолтный конструктор.

sonofR9 replied:
Здесь да, т.к. проверка до этого

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+        m_DescriptorByEndpoint.erase(endpoint);
+        m_endpointByDescriptor.erase(descriptor);
+
+        auto& docs = m_docsByUser[descriptor.login];
+        std::erase(docs, descriptor.documentId);
+
+        auto& users = m_usersByDoc[descriptor.documentId];
+        std::erase(users, descriptor.login);
```
Аналогичная проблема.

sonofR9 replied:
Если есть в одном, то гарантированно есть во всём остальном (добавляется и удаляется всегда всё одновременно)

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+    }
+}
+
+void InternalSessionsManager::RemoveSession(IServiceSession* session)
+{
+    if (!session)
+    {
+        return;
+    }
```
Ну вот да. А можно по ссылке передавать.

sonofR9 replied:
(дублирую)

Обычно они будут по указателям (умным зачастую) храниться, поэтому решил, что так извне будет удобнее пользоваться.
Если я не прав и всё равно удобнее по ссылке, то исправлю

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+        return;
+    }
+
+    const Endpoint& endpoint = session->GetClientEndpoint();
```
`const auto&`

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+
+    const Endpoint& endpoint = session->GetClientEndpoint();
+
+    if (m_DescriptorByEndpoint.contains(endpoint))
```
Опять инверсия.

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+    if (m_DescriptorByEndpoint.contains(endpoint))
+    {
+        const DocSessionDescriptor& descriptor = m_DescriptorByEndpoint[endpoint];
+        m_sessions.erase(endpoint);
+        m_endpointByDescriptor.erase(descriptor);
+        m_DescriptorByEndpoint.erase(endpoint);
+
+        auto& docs = m_docsByUser[descriptor.login];
+        std::erase(docs, descriptor.documentId);
+
+        auto& users = m_usersByDoc[descriptor.documentId];
+        std::erase(users, descriptor.login);
```
Вообще вот это вот все, кажется, выглядит один в один как то, что было в методе выше. Вынести в отдельный приватный метод?

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+std::weak_ptr<IServiceSession> InternalSessionsManager::GetSession(const DocSessionDescriptor& descriptor)
+{
+    std::weak_ptr<IServiceSession> result{};
+    if (m_endpointByDescriptor.contains(descriptor))
```
Инверсия.

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+
+std::weak_ptr<IServiceSession> InternalSessionsManager::GetSession(const Endpoint& endpoint)
+{
+    std::weak_ptr<IServiceSession> result{};
+    if (m_sessions.contains(endpoint))
+    {
+        result = m_sessions[endpoint];
+    }
+    return result;
```
```cpp
    if (m_sessions.contains(endpoint))
    {
        return {m_sessions[endpoint]};
    }
```

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+    std::vector<std::weak_ptr<IServiceSession>> sessions{};
+    if (m_usersByDoc.contains(doc))
+    {
+        for (const std::string& user : m_usersByDoc[doc])
```
`const auto&`

Arugaf commented on tests/backend/utils/base_service_chassis/internal_sessions_manager/test_internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,400 @@
+
+    EXPECT_EQ(actual.size(), expected.size());
+}
+} // namespace no_session
```
Ок.

Arugaf reviewed (changes requested):
Нужна читаемая ссылка на таску.

Нужно долить к себе запускаемый ci.

Часть комментариев к ревью распространяется на весь пр.

`[[nodiscard]]`, `noexcept`, `const` там, где это нужно.

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
INKLB-000 add const etc.

sonofR9 committed:
INKLB-000 REF constructor in return

sonofR9 committed:
INKLB-000 REF replaced duplicated code

sonofR9 committed:
Test ci: does not run

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
INKLB-000 FIX tests

sonofR9 committed:
FIX append run_test script function (needed ''')

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
INKLB-000 FIX include memory

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
INKLB-000 FIX remove session by ptr

Was getting reference to descriptor and then deleting everything using this reference.
And, obviousle, descriptor itself too

Arugaf commented on backend/utils/base_service_chassis/internal_sessions_manager/src/internal_sessions_manager.cpp:
```diff
@@ -0,0 +1,161 @@
+namespace
+{
+using IServiceSession = inklink::server_network::IServiceSession;
+}
```
nit: `// namespace`

sonofR9 committed:
Merge branch 'dev/a.novak/network' of github.com:cpp-park-vk-education/2023_1_The_Loopers into dev/a.novak/sessions_manager

sonofR9 committed:
INKLB-000 REF // namespace
